apps/champions/lib/champions/battle/simulator.ex

```diff
@@ -63,187 +64,465 @@ defmodule Champions.Battle.Simulator do
+      Enum.reduce(state.units, %{}, fn {unit_id, unit}, units ->
+        if unit.health > 0 do
+          Map.put(units, unit_id, unit)
+        else
+          Logger.info("Unit #{format_unit_name(unit)} died.")
+          units
+        end
+      end)
+
+    Map.put(state, :units, new_units)
+  end
+
+  # Check if the battle has ended.
+  # Battle can end if all unit of a team are dead, or if the maximum step amount has been reached.
+  defp check_winner(state, step) do
+    winner =
+      cond do
+        Enum.empty?(state.units) -> :tie
+        Enum.all?(state.units, fn {_id, unit} -> unit.team == 2 end) -> :team_2
+        Enum.all?(state.units, fn {_id, unit} -> unit.team == 1 end) -> :team_1
+        true -> :none
+      end
+
+    case winner do
+      :none ->
+        if step == @maximum_steps do
+          Logger.info("Battle timeout.")
+          {:halt, :timeout}
+        else
+          {:cont, state}
+        end
+
+      result ->
+        Logger.info("Battle ended. Result: #{result}.")
+        {:halt, result}
+    end
+  end
+
+  defp process_step_for_units(initial_step_state) do
+    Enum.reduce(initial_step_state.units, initial_step_state, fn {unit_id, unit}, current_state ->
+      Logger.info("Process step #{initial_step_state.step_number} for unit #{format_unit_name(unit)}")
+      process_step_for_unit(initial_step_state.units[unit_id], current_state, initial_step_state)
+    end)
+  end
+
+  # Calculate the new state of the battle after a step passes for a unit.
+  # Updates cooldowns, casts skills and reduces self-affecting modifier durations.
+  defp process_step_for_unit(unit, current_state, initial_step_state) do
     new_state =
       cond do
-        not can_attack(unit) ->
+        not can_attack(unit, initial_step_state) ->
+          Logger.info("Unit #{format_unit_name(unit)} cannot attack")
           current_state
 
         can_cast_ultimate_skill(unit) ->
-          unit
-          |> cast_skill(unit.ultimate_skill, initial_step_state, current_state)
-          |> put_in([unit.id, :energy], 0)
+          Logger.info("Unit #{format_unit_name(unit)} casting Ultimate skill")
+
+          current_state
+          |> Map.put(:skills_being_cast, [unit.ultimate_skill | current_state.skills_being_cast])
+          |> put_in([:units, unit.id, :energy], 0)
 
         can_cast_basic_skill(unit) ->
-          new_state = cast_skill(unit, unit.basic_skill, initial_step_state, current_state)
+          Logger.info("Unit #{format_unit_name(unit)} casting basic skill")
 
-          new_state
+          current_state
+          |> Map.put(:skills_being_cast, [unit.basic_skill | current_state.skills_being_cast])
           |> put_in(
-            [unit.id, :basic_skill, :remaining_cooldown],
+            [:units, unit.id, :basic_skill, :remaining_cooldown],
+            # We need this + 1 because we're going to reduce the cooldown at the end of the step
             unit.basic_skill.base_cooldown + 1
           )
```

Comment:
Can't you just avoid the cooldown decrease if the skill being cast is a basic, instead of doing this trick?

Reply:
mmh this is the same conversation that arose with @agustinesco [here](https://github.com/lambdaclass/mirra_backend/pull/373#discussion_r1541545956) 🤔

It's clearly a faulty point in the implementation but I think it's also the best option we currently have. If we were to undo the trick, we'd have to replicate the cooldown reduction process in every single other action we create, which doesn't scale nicely, so the trick attempts to fix just that.